secp256k1: configure with more feature flags

chris-martin opened the pull request:
###### Motivation for this change

Closes #18637
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

chris-martin committed:
secp256k1: 2016-05-30 -> 2016-11-27, more flags (#18637)

rasendubi commented on pkgs/tools/security/secp256k1/default.nix:
```diff
@@ -16,7 +16,12 @@ stdenv.mkDerivation rec {
 
-  configureFlags = [ "--enable-module-recovery" ];
+  configureFlags = [
+    "--enable-experimental"
```
Does it make sense to make these flags optional? I mean, when I see `experimental`, I doubt it should be enabled by default.

fpletz commented on pkgs/tools/security/secp256k1/default.nix:
```diff
@@ -16,7 +16,12 @@ stdenv.mkDerivation rec {
+    "--enable-module-schnorr"
+  ];
 
   preConfigure = "./autogen.sh";
```
You should instead use `autoreconfHook` in `nativeBuildInputs`.

rasendubi commented:
@chris-martin, are you still interested in the PR?

chris-martin commented:
Oh yeah, sorry, I'll update this in a bit.

rasendubi commented:
@chris-martin ping

rasendubi commented:
@chris-martin, another friendly reminder

chris-martin commented:
Sorry, I promise this is still on my to-do list!

rasendubi commented:
@chris-martin no worries – I understand you ;)

chris-martin commented:
Okay, updated!

* Bumped to latest version
* All flags are optional
* Added JNI flag
* Removed schorr flag (which apparently no longer exists)
* Using `autoreconfHook` as suggested

fpletz commented on pkgs/tools/security/secp256k1/default.nix:
```diff
@@ -1,32 +1,51 @@
+# which is a large dependency)
+, enableJNI ? false
+
+, ... }:
```
Small nitpick: `...` is not necessary here as `callPackage` will do the magic for you.

fpletz commented on pkgs/tools/security/secp256k1/default.nix:
```diff
@@ -1,32 +1,51 @@
+    optionals enableRecovery [ "--enable-module-recovery" ] ++
+    optionals enableJNI [ "--enable-jni" ];
 
   preConfigure = "./autogen.sh";
```
You can remove this hook because `autoreconfHook` already does this in a new build phase.

chris-martin commented:
Updated:

* Removed `, ...`
* Removed `preConfigure = "./autogen.sh";`

chris-martin commented:
I think this should be ready to go now.
